File: src/leetcode/python/medium/RemoveNthNodeFromEndOfList_19.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class RemoveNthNodeFromEndOfList_19(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        nth = fast = head
        for _ in range(n):
            fast = fast.next
        if not fast:
            return head.next
        while fast.next:
            fast = fast.next
            nth = nth.next
        nth.next = nth.next.next
        return head

File: src/leetcode/java/RemoveNthNodeFromEndOfList_19.java
package leetcode.java;

public class RemoveNthNodeFromEndOfList_19 {
  public ListNode removeNthFromEnd(ListNode head, int n) {
    ListNode fast = head;
    ListNode nth = head;
    int i = 0;
    while (i++ < n)
      fast = fast.next;
    if (fast == null) return head.next;
    while (fast.next != null) {
      fast = fast.next;
      nth = nth.next;
    }
    nth.next = nth.next.next;
    return head;
  }

  class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
  }
}
